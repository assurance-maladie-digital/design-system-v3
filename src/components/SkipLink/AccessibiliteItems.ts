import { ExpertiseLevelEnum } from './constants/ExpertiseLevelEnum'

export const AccessibiliteItemsIndeterminate = [
	{
		title: "Catégorie 6 : Liens",
		subtitle: ' 6.1 Chaque lien est-il explicite',
		items: [
			{

				precision: ' 6.1.1 Chaque lien texte vérifie-t-il une de ces conditions (hors cas particuliers) ? '+
    					'L’intitulé de lien seul permet d’en comprendre la fonction et la destination ; '+
    					'L’intitulé de lien additionné au contexte du lien permet d’en comprendre la fonction et la destination.',
				link: 'https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#6.1.1',
				solution: [{
					info1: '1. Retrouver dans le document les liens texte ;',
					info2: '2. Pour chaque lien texte, vérifier que ce qui permet d’en comprendre la fonction et la destination est : '+
    						' Soit l’intitulé du lien seul, soit le contexte du lien.' ,
					info3: '3. Si c’est le cas pour chaque lien texte, le test est validé.',
				}],
				expertise: ExpertiseLevelEnum.DEV,

			},

		],
	},
	{
		title: "Catégorie 10 : Présentation de l'information",
		subtitle: '10.8 Pour chaque page web, les contenus cachés ont-ils vocation à être ignorés par les technologies d’assistance',
		items: [
			{

				precision: '10.8.1 Le contenu caché a vocation à être ignoré par les technologies d’assistance , il n’a pas vocation à être ignoré par les technologies d’assistance et est rendu restituable par les technologies d’assistance suite à une action de l’utilisateur réalisable au clavier ou par tout dispositif de pointage sur un élément précédent le contenu caché ou suite à un repositionnement du focus dessus',
				link: 'https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#10.8.1',
				solution: [{
					info1: '1. Retrouver les contenus cachés (éléments pourvus de l’attribut hidden ou de l’attribut WAI-ARIA aria-hidden, ou bien d’une classe ou d’un ensemble de styles CSS susceptibles de masquer le contenu)',
					info2: '2. Pour chaque contenu caché, vérifier que :'+
    						'Soit le contenu caché a vocation à être ignoré par les technologies d’assistance (un élément statistique de visites par exemple) ;'+
    						'Soit le contenu caché n’a pas vocation à être ignoré par les technologies d’assistance, et dans ce cas il est rendu restituable par les technologies d’assistance au moyen :'+
        					'Soit d’une action de l’utilisateur réalisable au clavier ou par tout dispositif de pointage sur un élément précédent le contenu caché ;'+
        					'Soit d’une fonction de programmation qui repositionne le focus sur le contenu',
					info3: '3. Si c’est le cas pour chaque contenu caché, le test est validé',
				}],
				expertise: ExpertiseLevelEnum.DEV,

			},

		],
	},
	{
		title: "Catégorie 12 : Navigation",
		subtitle: '12.9 Dans chaque page web, la navigation ne doit pas contenir de piège au clavier. Cette règle est-elle respectée ?',
		items: [
			{

				precision: '12.9.1 Dans chaque page web, chaque élément recevant le focus vérifie-t-il une de ces conditions ? '+
				'Il est possible d’atteindre l’élément suivant ou précédent pouvant recevoir le focus avec la touche de tabulation ;'+
				'L’utilisateur est informé d’un mécanisme fonctionnel permettant d’atteindre au clavier l’élément suivant ou précédent pouvant recevoir le focus.',
				link: 'https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#12.9.1',
				solution: [{
					info1: '1. Retrouver dans le document l’ensemble des éléments d’interface susceptibles de recevoir le focus (au moyen de la tabulation ou au moyen d’un script) ;',
					info2: '2. Pour chaque élément d’interface, vérifier que l’utilisateur peut atteindre l’élément suivant ou précédent pouvant recevoir le focus : '+
    						'Soit au moyen de la touche de tabulation (Tab ou Maj+Tab) ;'+
        					'Soit au moyen d’une autre interaction clavier dont l’utilisateur est informé (par exemple, les flèches de direction).',
					info3: '3. Si c’est le cas pour chaque élément d’interface, le test est validé.',
				}],
				expertise: ExpertiseLevelEnum.DEV,

			},

		],
	},

]

export const AccessibiliteItemsValidated = [

	
	

]
