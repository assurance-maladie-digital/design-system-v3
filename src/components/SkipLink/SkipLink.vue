<script lang="ts" setup>
	import { locales } from './locales'
	import { nextTick, ref, onMounted } from 'vue'
	import { getCurrentInstance } from 'vue'
	import type { Router } from 'vue-router'

	withDefaults(
		defineProps<{
			label?: string
			target?: string
		}>(),
		{
			label: locales.label,
			target: '#main',
		},
	)

	const skipLinkSpan = ref<HTMLLinkElement | null>(null)

	onMounted(() => {
		const instance = getCurrentInstance()
		if (!instance) return

		// eslint-disable-next-line @typescript-eslint/no-explicit-any -- we test the existence of Nuxt
		const nuxtApp = (instance?.appContext.app as any)?.$nuxt
		let router: undefined | Router
		if (nuxtApp && nuxtApp.$router) {
			router = nuxtApp.$router as Router
		}

		// eslint-disable-next-line @typescript-eslint/no-explicit-any -- we test the existence of vue-router
		const vueRouter = (instance.appContext.app.config.globalProperties as any)?.$router
		if (vueRouter) {
			router = vueRouter as Router
		}

		if (router && router.afterEach) {
			router.afterEach((to, from, fail) => {
				if (fail) return
				if (to.path === from.path) return
				nextTick(() => {
					skipLinkSpan.value?.focus()
				})
			})
		}
	})
</script>

<template>
	<div class="vd-skip-link-container">
		<span
			ref="skipLinkSpan"
			tabindex="-1"
		/>

		<a
			:href="target"
			class="vd-skip-link text-primary d-block d-sr-only-focusable px-2"
		>
			<slot>{{ label }}</slot>
		</a>
	</div>
</template>

<style lang="scss" scoped>
@use '/src/assets/tokens.scss';

.vd-skip-link {
	z-index: 150;
	position: fixed;
	top: 0;
	right: 0;
	transition: none;
	width: 100%;
	background: #fff;
	outline: none; // Disable outline to use border
	border: 2px solid tokens.$blue-darken-60;
}
</style>
