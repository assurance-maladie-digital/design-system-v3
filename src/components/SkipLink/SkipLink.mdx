import {Controls, Canvas, Meta, Source} from '@storybook/blocks';
import SkipLink from './SkipLink.vue';

import * as SkipLinkStories from './SkipLink.stories';

<Meta of={SkipLinkStories} />

# SkipLink

Le composant `SkipLink` est utilisé pour permettre à l'utilisateur utilisant la navigation au clavier de passer directement à une section de contenu.

Il est généralement placé en haut de la page et est caché visuellement. Lorsqu'il est activé, il devient visible et permet à l'utilisateur de sauter directement à une section de contenu.

Quand l'utilisateur change de page ou de vue, le focus est automatiquement placé sur l'élément de saut pour lui permettre de naviguer rapidement dans la page.

<Canvas of={SkipLinkStories.Default}  />

## API

<Controls of={SkipLinkStories.Default} />

## Exemple d'utilisation avec Nuxt

<Source dark code={`
// layouts/default.vue
<template>
  <div>
    <SkipLink label="Aller au contenu" target="#main-content"/>
    <div id="main-content">
      <slot />
    </div>
  </div>
</template>
`} />

## Exemple d'utilisation avec Vue

<Source dark code={`
// main.vue
<template>
  <v-app>
    <SkipLink target="#main-content"/>
    <nav>
      <ul>
        <li>
          <router-link to="/">index</router-link>
        </li>
        <li>
          <router-link to="/other">other</router-link>
        </li>
      </ul>
    </nav>
    <div id="main-content">
      <router-view />
    </div>
  </v-app>
</template>
`} />
