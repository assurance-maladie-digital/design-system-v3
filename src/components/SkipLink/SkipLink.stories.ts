import type { Meta, StoryObj } from '@storybook/vue3'
import SkipLink from './SkipLink.vue'

const meta = {
	title: 'Components/SkipLink',
	component: SkipLink,
	argTypes: {
		default: {
			control: { type: 'text' },
			default: 'Skip to content',
		},
	},
	parameters: {
		layout: 'fullscreen',
	},
} as Meta<typeof SkipLink>

export default meta

type Story = StoryObj<typeof meta>

export const Default: Story = {
	args: {
		target: 'http://localhost:6007/?path=/docs/components-skiplink--docs#main',
	},
	render: (args) => {
		return {
			components: { SkipLink },
			setup() {
				return { args }
			},
			template: `
				<div class="pa-8">
					<p>Pour afficher le composant, cliquez ici et appuyer sur <kbd>Tab</kbd>.</p>
					<SkipLink 
						:target="args.target"
						:label="args.label"
					>
						<template #default v-if="args.default"><span v-html="args.default"/></template>
					</SkipLink>
				</div>
			`,
		}
	},
}

export const WithSlot: Story = {
	args: {
		target: 'http://localhost:6007/?path=/docs/components-skiplink--docs#main',
		default: '<b>lorem ipsum</b>',
	},
	render: (args) => {
		return {
			components: { SkipLink },
			setup() {
				return { args }
			},
			template: `
				<div class="pa-8">
					<p>Pour afficher le composant, cliquez ici et appuyer sur <kbd>Tab</kbd>.</p>
					<SkipLink 
						v-bind="args"
					>
						<template #default v-if="args.default"><span v-html="args.default"/></template>
					</SkipLink>
				</div>
			`,
		}
	},
}
