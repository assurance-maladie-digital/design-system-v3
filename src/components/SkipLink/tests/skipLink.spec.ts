import { describe, it, expect, beforeEach, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import { createRouter, createWebHistory } from 'vue-router';
import SkipLink from '../SkipLink.vue';

// Create a mock router
const router = createRouter({
	history: createWebHistory(),
	routes: [
		{ path: '/', component: { template: '<div>Home</div>' } },
		{ path: '/about', component: { template: '<div>About</div>' } },
	],
});

describe('SkipLink', () => {
	beforeEach(() => {
		vi.restoreAllMocks();
	});

	it('renders correctly', async () => {
		const wrapper = mount(SkipLink, {
			global: {
				plugins: [router],
			},
		});

		expect(wrapper.html()).toMatchSnapshot();
	});

	it('focuses the skip link on route change', async () => {
		const wrapper = mount(SkipLink, {
			global: {
				plugins: [router],
			},
		});

		// Create a mock for the focus method
		const spy = vi.spyOn(wrapper.vm.$refs.skipLinkSpan as HTMLLinkElement, 'focus');

		// Trigger the route change
		await router.push('/about');
		await router.isReady(); // Ensure the router is ready

		expect(spy).toHaveBeenCalled();
	});
});