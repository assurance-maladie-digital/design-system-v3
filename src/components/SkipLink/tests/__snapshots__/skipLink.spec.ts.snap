// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SkipLink > renders correctly 1`] = `"<div data-v-bf1925a1="" class="vd-skip-link-container"><span data-v-bf1925a1="" tabindex="-1"></span><a data-v-bf1925a1="" href="#main" class="vd-skip-link text-primary d-block d-sr-only-focusable px-2">Aller au contenu principal</a></div>"`;
